In [1]:
import numpy as np
from hmmlearn import hmm
import pandas as pd
import math
import warnings
warnings.filterwarnings('ignore')

# Dataset

We decided to use the text from a book in order to predict and sample words using Hidden Markov Models. 

We download the data from the Project Gutenberg (http://www.gutenberg.org/).
Project Gutenberg offers over 58,000 free eBooks. Project Gutenberg was the first provider of free electronic books, or eBooks. 

We selected the following book:
"The Man in the Iron Mask" by Alexandre Dumas

In [2]:
filename = 'C:/Users/Caio Laptop/OneDrive - The University of Kansas/Documents/PhD/11. Courses/22. EECS 738 - Machine Learning/' \
            'Assignments/2. Hiden Markov Models/The_Man_in_the_Iron_Mask_by _Alexandre _Dumas3.txt'  

filename_c1 = 'C:/Users/Caio Laptop/OneDrive - The University of Kansas/Documents/PhD/11. Courses/22. EECS 738 - Machine Learning/' \
            'Assignments/2. Hiden Markov Models/The_Man_in_the_Iron_Mask_by _Alexandre _Dumas_chapter1.txt'

filename_c2 = 'C:/Users/Caio Laptop/OneDrive - The University of Kansas/Documents/PhD/11. Courses/22. EECS 738 - Machine Learning/' \
            'Assignments/2. Hiden Markov Models/The_Man_in_the_Iron_Mask_by _Alexandre _Dumas_chapter2.txt'

# Dictionaries and probabilities of Word Occurrence

The first file has the entire book text. Since it is a long a text, we decided to focus our analysis in the first 2 chapters.

The next two file has the text corpus of Chapter I and Chapter II of the book.

Now we defined some loops to go over each of the files, break the text in lines and create a separate dictionary of the used words in each file.

Thus, we will have the following dictionaries:

* new_dict
* new_dict1
* new_dict2

In [3]:
new_dict ={}
total_count = 0
with open(filename, "r") as fp:   
    for line in fp:     
        words = line.split()
        for word in words:
            word = word.lower()        
        if not word in new_dict:
             new_dict[word] = 1
        else:
            new_dict[word] +=1
            total_count +=1

new_dict1 ={}
total_count = 0
with open(filename_c1, "r") as fp:   
    for line in fp:     
        words = line.split()
        for word in words:
            word = word.lower()        
        if not word in new_dict1:
             new_dict1[word] = 1
        else:
            new_dict1[word] +=1
            total_count +=1

new_dict2 ={}
total_count = 0            
with open(filename_c2, "r") as fp:   
    for line in fp:     
        words = line.split()
        for word in words:
            word = word.lower()        
        if not word in new_dict2:
             new_dict2[word] = 1
        else:
            new_dict2[word] +=1
            total_count +=1

Let's take a quick look on our 3 dictionaries.

In [4]:
new_dict

{'mask': 2,
 'dumas': 2,
 'introduction': 2,
 'the': 923,
 'playwright': 1,
 'had': 88,
 'a': 226,
 'adventures': 1,
 'almost': 7,
 'and': 308,
 'readers': 1,
 'friends': 48,
 'english': 4,
 'history': 5,
 'form': 2,
 'brief': 1,
 'novels': 2,
 '1625': 1,
 'athos': 45,
 'cardinal’s': 1,
 'battle': 1,
 'landlord': 1,
 'them': 88,
 'spy': 3,
 'of': 404,
 'now': 41,
 'has': 32,
 'sit': 1,
 'xiv': 19,
 'husband': 5,
 'have': 75,
 'de': 43,
 'bragelonne': 16,
 'but': 49,
 'at': 77,
 'brings': 1,
 'monarch': 2,
 'death': 26,
 'our': 17,
 'mazarin': 6,
 'october': 1,
 'does': 8,
 'edition': 4,
 'later': 4,
 'three': 6,
 'etexts': 2,
 'become': 6,
 'with': 126,
 'on': 52,
 'england': 5,
 'fortune': 6,
 'rich': 3,
 'king’s': 18,
 'own': 21,
 'to': 342,
 'formerly': 7,
 'fouquet': 167,
 'any': 21,
 'intendant': 7,
 'fouquet’s': 6,
 'attention': 13,
 'could': 16,
 'operation': 3,
 'sends': 1,
 'his': 205,
 'fortifications': 4,
 'all': 79,
 'handwriting': 1,
 'aramis': 179,
 'vannes': 23,
 '

In [5]:
new_dict1

{'1625': 1,
 '_ennui_': 1,
 '_fetes_': 1,
 '_too_': 1,
 'a': 18,
 'about': 1,
 'absence': 1,
 'accident': 1,
 'act': 2,
 'adamant': 2,
 'adventures': 1,
 'affairs': 1,
 'affrighted': 1,
 'after': 2,
 'again': 2,
 'ago': 2,
 'ah': 1,
 'alas': 2,
 'all': 11,
 'almost': 2,
 'along': 2,
 'already': 1,
 'also': 1,
 'alteration': 2,
 'always': 2,
 'am': 4,
 'ambitious': 1,
 'an': 3,
 'and': 29,
 'answer': 4,
 'any': 4,
 'aramis': 30,
 'are': 2,
 'armchair': 3,
 'arms': 2,
 'arrest': 1,
 'arriving': 1,
 'ask': 3,
 'asked': 3,
 'asleep': 1,
 'aspect': 2,
 'assassinated': 2,
 'assuredly': 1,
 'astonished': 1,
 'at': 7,
 'athos': 1,
 'attached': 1,
 'attachment': 1,
 'attention': 3,
 'attitude': 2,
 'austria': 2,
 'austria--': 2,
 'aware': 1,
 'back': 1,
 'baisemeaux': 2,
 'barrier': 1,
 'bars': 2,
 'base;': 1,
 'bastile': 9,
 'battle': 1,
 'be': 4,
 'become': 1,
 'been': 2,
 'before': 3,
 'being': 2,
 'beneath': 1,
 'beseech': 1,
 'better': 7,
 'beyond': 1,
 'bid': 1,
 'bishop': 4,
 'bitter': 1

In [6]:
new_dict1

{'1625': 1,
 '_ennui_': 1,
 '_fetes_': 1,
 '_too_': 1,
 'a': 18,
 'about': 1,
 'absence': 1,
 'accident': 1,
 'act': 2,
 'adamant': 2,
 'adventures': 1,
 'affairs': 1,
 'affrighted': 1,
 'after': 2,
 'again': 2,
 'ago': 2,
 'ah': 1,
 'alas': 2,
 'all': 11,
 'almost': 2,
 'along': 2,
 'already': 1,
 'also': 1,
 'alteration': 2,
 'always': 2,
 'am': 4,
 'ambitious': 1,
 'an': 3,
 'and': 29,
 'answer': 4,
 'any': 4,
 'aramis': 30,
 'are': 2,
 'armchair': 3,
 'arms': 2,
 'arrest': 1,
 'arriving': 1,
 'ask': 3,
 'asked': 3,
 'asleep': 1,
 'aspect': 2,
 'assassinated': 2,
 'assuredly': 1,
 'astonished': 1,
 'at': 7,
 'athos': 1,
 'attached': 1,
 'attachment': 1,
 'attention': 3,
 'attitude': 2,
 'austria': 2,
 'austria--': 2,
 'aware': 1,
 'back': 1,
 'baisemeaux': 2,
 'barrier': 1,
 'bars': 2,
 'base;': 1,
 'bastile': 9,
 'battle': 1,
 'be': 4,
 'become': 1,
 'been': 2,
 'before': 3,
 'being': 2,
 'beneath': 1,
 'beseech': 1,
 'better': 7,
 'beyond': 1,
 'bid': 1,
 'bishop': 4,
 'bitter': 1

With these 3 dictionaries, now we can easily calculate the probabilities of occurrence o each word in each one the dictionaries.

In order to do that, we will create the following dictionaries:

* words_probabilities
* words_probabilities1
* words_probabilities2

In [7]:
my_new_dict = pd.DataFrame.from_dict(new_dict, orient='index')
words_probabilities={}
for word in new_dict:            
     probab = new_dict[word]/ my_new_dict.sum()
     words_probabilities[word]=probab
     print('probability of word '+ word +': ' + str(probab))


probability of word mask: 0    0.000089
dtype: float64
probability of word dumas: 0    0.000089
dtype: float64
probability of word introduction: 0    0.000089
dtype: float64
probability of word the: 0    0.040969
dtype: float64
probability of word playwright: 0    0.000044
dtype: float64
probability of word had: 0    0.003906
dtype: float64
probability of word a: 0    0.010032
dtype: float64
probability of word adventures: 0    0.000044
dtype: float64
probability of word almost: 0    0.000311
dtype: float64
probability of word and: 0    0.013671
dtype: float64
probability of word readers: 0    0.000044
dtype: float64
probability of word friends: 0    0.002131
dtype: float64
probability of word english: 0    0.000178
dtype: float64
probability of word history: 0    0.000222
dtype: float64
probability of word form: 0    0.000089
dtype: float64
probability of word brief: 0    0.000044
dtype: float64
probability of word novels: 0    0.000089
dtype: float64
probability of word 1625: 0    0.

probability of word prisoner: 0    0.001909
dtype: float64
probability of word order: 0    0.001997
dtype: float64
probability of word place: 0    0.000932
dtype: float64
probability of word was: 0    0.005992
dtype: float64
probability of word said: 0    0.003684
dtype: float64
probability of word merely: 0    0.000089
dtype: float64
probability of word resounded: 0    0.000044
dtype: float64
probability of word towers: 0    0.000044
dtype: float64
probability of word luxury: 0    0.000044
dtype: float64
probability of word effected: 0    0.000133
dtype: float64
probability of word curious: 0    0.000089
dtype: float64
probability of word were: 0    0.001154
dtype: float64
probability of word arriving: 0    0.000089
dtype: float64
probability of word prisoner’s: 0    0.000089
dtype: float64
probability of word do: 0    0.002131
dtype: float64
probability of word confession: 0    0.000089
dtype: float64
probability of word lantern: 0    0.000178
dtype: float64
probability of word alr

probability of word even: 0    0.000888
dtype: float64
probability of word year--: 0    0.000089
dtype: float64
probability of word ago: 0    0.000178
dtype: float64
probability of word time: 0    0.00142
dtype: float64
probability of word work: 0    0.000577
dtype: float64
probability of word world: 0    0.000311
dtype: float64
probability of word long: 0    0.000533
dtype: float64
probability of word over: 0    0.000533
dtype: float64
probability of word ‘perronnette: 0    0.000044
dtype: float64
probability of word called: 0    0.000133
dtype: float64
probability of word hastily: 0    0.000044
dtype: float64
probability of word windows: 0    0.000044
dtype: float64
probability of word brim: 0    0.000044
dtype: float64
probability of word affrighted: 0    0.000044
dtype: float64
probability of word hear: 0    0.000178
dtype: float64
probability of word went: 0    0.000222
dtype: float64
probability of word edge;: 0    0.000044
dtype: float64
probability of word misfortune’: 0   

probability of word forever: 0    0.000488
dtype: float64
probability of word we: 0    0.000932
dtype: float64
probability of word jailer: 0    0.000133
dtype: float64
probability of word trusty: 0    0.000044
dtype: float64
probability of word every: 0    0.000444
dtype: float64
probability of word barrier: 0    0.000044
dtype: float64
probability of word break: 0    0.000178
dtype: float64
probability of word bars: 0    0.000133
dtype: float64
probability of word note: 0    0.000178
dtype: float64
probability of word manner: 0    0.001154
dtype: float64
probability of word those: 0    0.000799
dtype: float64
probability of word base;: 0    0.000044
dtype: float64
probability of word plain: 0    0.000222
dtype: float64
probability of word friend: 0    0.004572
dtype: float64
probability of word moment’s: 0    0.000089
dtype: float64
probability of word mind: 0    0.000577
dtype: float64
probability of word pressing: 0    0.000133
dtype: float64
probability of word need: 0    0.00013

probability of word look: 0    0.001198
dtype: float64
probability of word see--: 0    0.000178
dtype: float64
probability of word halles: 0    0.000178
dtype: float64
probability of word or: 0    0.001642
dtype: float64
probability of word quarters: 0    0.000089
dtype: float64
probability of word proud: 0    0.000089
dtype: float64
probability of word tailor’s: 0    0.000089
dtype: float64
probability of word doors: 0    0.000044
dtype: float64
probability of word whom: 0    0.000799
dtype: float64
probability of word costumes: 0    0.000044
dtype: float64
probability of word five: 0    0.000355
dtype: float64
probability of word contented: 0    0.000044
dtype: float64
probability of word insisted: 0    0.000089
dtype: float64
probability of word ribbons: 0    0.000044
dtype: float64
probability of word unless: 0    0.000044
dtype: float64
probability of word counter: 0    0.000089
dtype: float64
probability of word pronounced: 0    0.000178
dtype: float64
probability of word frien

probability of word glass;’: 0    0.000044
dtype: float64
probability of word (for: 0    0.000044
dtype: float64
probability of word soft: 0    0.000089
dtype: float64
probability of word reflection’: 0    0.000089
dtype: float64
probability of word figure: 0    0.000178
dtype: float64
probability of word way: 0    0.001198
dtype: float64
probability of word similar: 0    0.000044
dtype: float64
probability of word juxtaposition: 0    0.000089
dtype: float64
probability of word vocabulary: 0    0.000089
dtype: float64
probability of word strategic: 0    0.000044
dtype: float64
probability of word experiments: 0    0.000089
dtype: float64
probability of word knocked: 0    0.000044
dtype: float64
probability of word body: 0    0.000133
dtype: float64
probability of word voliere--: 0    0.000266
dtype: float64
probability of word chalk: 0    0.000044
dtype: float64
probability of word wearer’: 0    0.000089
dtype: float64
probability of word practice: 0    0.000133
dtype: float64
pr

probability of word money: 0    0.000266
dtype: float64
probability of word departing: 0    0.000089
dtype: float64
probability of word business: 0    0.000355
dtype: float64
probability of word millions: 0    0.000178
dtype: float64
probability of word superintendent: 0    0.000444
dtype: float64
probability of word fall--: 0    0.000089
dtype: float64
probability of word ‘fall’: 0    0.000089
dtype: float64
probability of word represent: 0    0.000089
dtype: float64
probability of word poets: 0    0.000089
dtype: float64
probability of word elegant: 0    0.000089
dtype: float64
probability of word people: 0    0.000311
dtype: float64
probability of word friendship: 0    0.000222
dtype: float64
probability of word prelate: 0    0.000089
dtype: float64
probability of word school: 0    0.000089
dtype: float64
probability of word cachet_: 0    0.000089
dtype: float64
probability of word jesuits: 0    0.000178
dtype: float64
probability of word crime: 0    0.000577
dtype: float64
prob

dtype: float64
probability of word near: 0    0.000222
dtype: float64
probability of word intend: 0    0.000044
dtype: float64
probability of word going: 0    0.000444
dtype: float64
probability of word whatever: 0    0.000133
dtype: float64
probability of word arm: 0    0.000178
dtype: float64
probability of word search: 0    0.000178
dtype: float64
probability of word courtyard: 0    0.000044
dtype: float64
probability of word sealed: 0    0.000044
dtype: float64
probability of word cleared: 0    0.000044
dtype: float64
probability of word left;: 0    0.000044
dtype: float64
probability of word horses: 0    0.000355
dtype: float64
probability of word villeneuve: 0    0.000044
dtype: float64
probability of word sign: 0    0.000222
dtype: float64
probability of word dream: 0    0.000311
dtype: float64
probability of word converse: 0    0.000089
dtype: float64
probability of word postilion: 0    0.000089
dtype: float64
probability of word taken: 0    0.000577
dtype: float64
probability 

probability of word saint-aignan: 0    0.000222
dtype: float64
probability of word latter: 0    0.000577
dtype: float64
probability of word wrong: 0    0.000133
dtype: float64
probability of word friend’: 0    0.000089
dtype: float64
probability of word monk: 0    0.000044
dtype: float64
probability of word served: 0    0.000044
dtype: float64
probability of word owes: 0    0.000044
dtype: float64
probability of word seem: 0    0.000133
dtype: float64
probability of word safe: 0    0.000089
dtype: float64
probability of word d’herblay;: 0    0.000044
dtype: float64
probability of word french: 0    0.000222
dtype: float64
probability of word present: 0    0.000444
dtype: float64
probability of word state: 0    0.000266
dtype: float64
probability of word medici: 0    0.000044
dtype: float64
probability of word delight: 0    0.000266
dtype: float64
probability of word queen’s: 0    0.000044
dtype: float64
probability of word labors: 0    0.000044
dtype: float64
probability of word a

probability of word me--: 0    0.000533
dtype: float64
probability of word suspicion: 0    0.000133
dtype: float64
probability of word foot: 0    0.000266
dtype: float64
probability of word project: 0    0.000799
dtype: float64
probability of word nonsense: 0    0.000089
dtype: float64
probability of word revealed: 0    0.000089
dtype: float64
probability of word arrives: 0    0.000089
dtype: float64
probability of word sake: 0    0.000178
dtype: float64
probability of word nobly: 0    0.000178
dtype: float64
probability of word speaking: 0    0.000178
dtype: float64
probability of word me;: 0    0.000133
dtype: float64
probability of word old: 0    0.000311
dtype: float64
probability of word times: 0    0.000133
dtype: float64
probability of word bible: 0    0.000044
dtype: float64
probability of word account: 0    0.000222
dtype: float64
probability of word pronounce: 0    0.000044
dtype: float64
probability of word feeling: 0    0.000222
dtype: float64
probability of word enough: 0 

probability of word throughout: 0    0.000044
dtype: float64
probability of word gave: 0    0.000222
dtype: float64
probability of word swiss: 0    0.000089
dtype: float64
probability of word step: 0    0.000178
dtype: float64
probability of word wounded: 0    0.000178
dtype: float64
probability of word quick: 0    0.000178
dtype: float64
probability of word directions: 0    0.000089
dtype: float64
probability of word insufficient: 0    0.000044
dtype: float64
probability of word anger: 0    0.000399
dtype: float64
probability of word changes: 0    0.000089
dtype: float64
probability of word unfortunately: 0    0.000044
dtype: float64
probability of word impertinently: 0    0.000089
dtype: float64
probability of word done;: 0    0.000044
dtype: float64
probability of word positively: 0    0.000044
dtype: float64
probability of word send: 0    0.000133
dtype: float64
probability of word harshness: 0    0.000044
dtype: float64
probability of word termination: 0    0.000044
dtype: float64

dtype: float64
probability of word knife: 0    0.000089
dtype: float64
probability of word hours: 0    0.000266
dtype: float64
probability of word turnkeys: 0    0.000044
dtype: float64
probability of word canopy: 0    0.000044
dtype: float64
probability of word became: 0    0.000133
dtype: float64
probability of word altogether: 0    0.000089
dtype: float64
probability of word result: 0    0.000133
dtype: float64
probability of word ordinary: 0    0.000133
dtype: float64
probability of word disturbance: 0    0.000044
dtype: float64
probability of word mere: 0    0.000044
dtype: float64
probability of word commendable: 0    0.000044
dtype: float64
probability of word breakfast: 0    0.000266
dtype: float64
probability of word adopt: 0    0.000044
dtype: float64
probability of word action: 0    0.000266
dtype: float64
probability of word disgrace;: 0    0.000044
dtype: float64
probability of word doubt: 0    0.000311
dtype: float64
probability of word real: 0    0.000044
dtype: float64


probability of word warning--: 0    0.000089
dtype: float64
probability of word graciousness: 0    0.000044
dtype: float64
probability of word transmit: 0    0.000044
dtype: float64
probability of word monarch’s: 0    0.000044
dtype: float64
probability of word unmoved: 0    0.000178
dtype: float64
probability of word le: 0    0.000399
dtype: float64
probability of word obtained: 0    0.000133
dtype: float64
probability of word conducted: 0    0.000133
dtype: float64
probability of word preserving: 0    0.000044
dtype: float64
probability of word tones: 0    0.000089
dtype: float64
probability of word instinctive: 0    0.000089
dtype: float64
probability of word eager: 0    0.000044
dtype: float64
probability of word making: 0    0.000178
dtype: float64
probability of word complaint: 0    0.000044
dtype: float64
probability of word course: 0    0.000089
dtype: float64
probability of word robber: 0    0.000089
dtype: float64
probability of word suborner: 0    0.000089
dtype: float64
p

probability of word steps: 0    0.000488
dtype: float64
probability of word armed: 0    0.000089
dtype: float64
probability of word shouted: 0    0.000044
dtype: float64
probability of word horror: 0    0.000089
dtype: float64
probability of word sweat: 0    0.000089
dtype: float64
probability of word vengeance: 0    0.000089
dtype: float64
probability of word tenderness: 0    0.000089
dtype: float64
probability of word suffered: 0    0.000089
dtype: float64
probability of word ashamed: 0    0.000044
dtype: float64
probability of word movement;: 0    0.000044
dtype: float64
probability of word forgive: 0    0.000044
dtype: float64
probability of word weakness: 0    0.000355
dtype: float64
probability of word mind;: 0    0.000044
dtype: float64
probability of word struck: 0    0.000222
dtype: float64
probability of word twin: 0    0.000044
dtype: float64
probability of word criminals: 0    0.000089
dtype: float64
probability of word brother--: 0    0.000089
dtype: float64
probability of

probability of word due: 0    0.000044
dtype: float64
probability of word narrow: 0    0.000044
dtype: float64
probability of word predict: 0    0.000044
dtype: float64
probability of word raoul;: 0    0.000089
dtype: float64
probability of word adieux: 0    0.000089
dtype: float64
probability of word comte: 0    0.000533
dtype: float64
probability of word hinted: 0    0.000089
dtype: float64
probability of word _brevet_: 0    0.000089
dtype: float64
probability of word porthos’s: 0    0.000044
dtype: float64
probability of word pursued: 0    0.000089
dtype: float64
probability of word complacently: 0    0.000089
dtype: float64
probability of word error: 0    0.000089
dtype: float64
probability of word nevertheless: 0    0.000311
dtype: float64
probability of word majeste_: 0    0.000089
dtype: float64
probability of word wheels: 0    0.000089
dtype: float64
probability of word acting: 0    0.000089
dtype: float64
probability of word whither: 0    0.000222
dtype: float64
probability 

probability of word conspicuous: 0    0.000044
dtype: float64
probability of word profession: 0    0.000044
dtype: float64
probability of word buy: 0    0.000044
dtype: float64
probability of word acquisition: 0    0.000089
dtype: float64
probability of word aimed: 0    0.000044
dtype: float64
probability of word facetiousness: 0    0.000089
dtype: float64
probability of word embarrassed: 0    0.000133
dtype: float64
probability of word hesitating: 0    0.000089
dtype: float64
probability of word tradesman: 0    0.000044
dtype: float64
probability of word planchet: 0    0.000178
dtype: float64
probability of word upstairs--: 0    0.000089
dtype: float64
probability of word rudeness: 0    0.000089
dtype: float64
probability of word time--: 0    0.000089
dtype: float64
probability of word respects: 0    0.000089
dtype: float64
probability of word forced: 0    0.000044
dtype: float64
probability of word idyls: 0    0.000044
dtype: float64
probability of word longus: 0    0.000044
dtype: f

probability of word amuse: 0    0.000044
dtype: float64
probability of word ourselves: 0    0.000178
dtype: float64
probability of word freely: 0    0.000089
dtype: float64
probability of word fort: 0    0.000355
dtype: float64
probability of word lerin: 0    0.000044
dtype: float64
probability of word correctly: 0    0.000089
dtype: float64
probability of word habit: 0    0.000089
dtype: float64
probability of word forms: 0    0.000044
dtype: float64
probability of word resists: 0    0.000089
dtype: float64
probability of word die: 0    0.000311
dtype: float64
probability of word presence--oh: 0    0.000044
dtype: float64
probability of word sees: 0    0.000044
dtype: float64
probability of word nowhere: 0    0.000044
dtype: float64
probability of word weakness;: 0    0.000044
dtype: float64
probability of word succeed: 0    0.000133
dtype: float64
probability of word cutting: 0    0.000044
dtype: float64
probability of word honest: 0    0.000044
dtype: float64
probability of word fas

dtype: float64
probability of word protector: 0    0.000044
dtype: float64
probability of word vulgar: 0    0.000044
dtype: float64
probability of word infallible: 0    0.000044
dtype: float64
probability of word receive: 0    0.000133
dtype: float64
probability of word overcast: 0    0.000089
dtype: float64
probability of word belliere: 0    0.000266
dtype: float64
probability of word unwell: 0    0.000133
dtype: float64
probability of word sacrilege: 0    0.000089
dtype: float64
probability of word poet: 0    0.000089
dtype: float64
probability of word happen_: 0    0.000089
dtype: float64
probability of word laugh: 0    0.000266
dtype: float64
probability of word afraid: 0    0.000178
dtype: float64
probability of word driven: 0    0.000044
dtype: float64
probability of word eyrie: 0    0.000089
dtype: float64
probability of word enthusiasm: 0    0.000089
dtype: float64
probability of word courier: 0    0.000044
dtype: float64
probability of word ceremonies: 0    0.000089
dtype: flo

probability of word warmly: 0    0.000133
dtype: float64
probability of word inquiries: 0    0.000133
dtype: float64
probability of word times--: 0    0.000089
dtype: float64
probability of word accused: 0    0.000178
dtype: float64
probability of word indeed;: 0    0.000044
dtype: float64
probability of word states: 0    0.000178
dtype: float64
probability of word becomes: 0    0.000044
dtype: float64
probability of word allowed: 0    0.000044
dtype: float64
probability of word health: 0    0.000089
dtype: float64
probability of word belle-isle’: 0    0.000089
dtype: float64
probability of word seducing: 0    0.000044
dtype: float64
probability of word pretty: 0    0.000044
dtype: float64
probability of word pleases: 0    0.000089
dtype: float64
probability of word ringing: 0    0.000089
dtype: float64
probability of word bedclothes: 0    0.000089
dtype: float64
probability of word footman: 0    0.000044
dtype: float64
probability of word beg: 0    0.000133
dtype: float64
probabilit

dtype: float64
probability of word flag: 0    0.000178
dtype: float64
probability of word troops: 0    0.000089
dtype: float64
probability of word probably: 0    0.000178
dtype: float64
probability of word sounded: 0    0.000133
dtype: float64
probability of word batteries: 0    0.000089
dtype: float64
probability of word blandest: 0    0.000044
dtype: float64
probability of word rolled;: 0    0.000044
dtype: float64
probability of word dikes: 0    0.000044
dtype: float64
probability of word matches: 0    0.000044
dtype: float64
probability of word lieutenant: 0    0.000178
dtype: float64
probability of word belong: 0    0.000089
dtype: float64
probability of word insinuation: 0    0.000089
dtype: float64
probability of word skiff: 0    0.000044
dtype: float64
probability of word barks: 0    0.000044
dtype: float64
probability of word soldiers: 0    0.000089
dtype: float64
probability of word rounds: 0    0.000089
dtype: float64
probability of word comrades: 0    0.000089
dtype: float6

probability of word children: 0    0.000089
dtype: float64
probability of word betwixt: 0    0.000044
dtype: float64
probability of word forget;: 0    0.000044
dtype: float64
probability of word hosts: 0    0.000089
dtype: float64
probability of word singular: 0    0.000044
dtype: float64
probability of word warn: 0    0.000044
dtype: float64
probability of word blunder: 0    0.000044
dtype: float64
probability of word formal: 0    0.000133
dtype: float64
probability of word crowd: 0    0.000044
dtype: float64
probability of word cavern: 0    0.000666
dtype: float64
probability of word us--_yet_: 0    0.000178
dtype: float64
probability of word locmaria: 0    0.000089
dtype: float64
probability of word mass: 0    0.000044
dtype: float64
probability of word frequently: 0    0.000044
dtype: float64
probability of word season: 0    0.000089
dtype: float64
probability of word afresh: 0    0.000089
dtype: float64
probability of word cautiously: 0    0.000044
dtype: float64
probability of wo

probability of word formidable: 0    0.000044
dtype: float64
probability of word cried;: 0    0.000044
dtype: float64
probability of word quickly: 0    0.000133
dtype: float64
probability of word portion: 0    0.000044
dtype: float64
probability of word tomb: 0    0.000222
dtype: float64
probability of word fail: 0    0.000133
dtype: float64
probability of word angel: 0    0.000044
dtype: float64
probability of word groan--he: 0    0.000044
dtype: float64
probability of word believed: 0    0.000089
dtype: float64
probability of word load: 0    0.000044
dtype: float64
probability of word collapse: 0    0.000089
dtype: float64
probability of word weaker: 0    0.000044
dtype: float64
probability of word triple: 0    0.000044
dtype: float64
probability of word epitaph: 0    0.000089
dtype: float64
probability of word oars: 0    0.000044
dtype: float64
probability of word receded: 0    0.000044
dtype: float64
probability of word drove: 0    0.000044
dtype: float64
probability of word carryi

probability of word chambre_: 0    0.000089
dtype: float64
probability of word torture: 0    0.000044
dtype: float64
probability of word dreaming: 0    0.000089
dtype: float64
probability of word dislodged: 0    0.000044
dtype: float64
probability of word blois: 0    0.000044
dtype: float64
probability of word imagination: 0    0.000044
dtype: float64
probability of word among: 0    0.000044
dtype: float64
probability of word degrees: 0    0.000044
dtype: float64
probability of word supplicating: 0    0.000044
dtype: float64
probability of word match: 0    0.000044
dtype: float64
probability of word firmament: 0    0.000044
dtype: float64
probability of word understanding: 0    0.000044
dtype: float64
probability of word diamonds: 0    0.000044
dtype: float64
probability of word picardy: 0    0.000044
dtype: float64
probability of word stiff: 0    0.000044
dtype: float64
probability of word shores: 0    0.000044
dtype: float64
probability of word express: 0    0.000044
dtype: float64
p

probability of word whence: 0    0.000044
dtype: float64
probability of word hat: 0    0.000044
dtype: float64
probability of word rebel: 0    0.000044
dtype: float64
probability of word desired: 0    0.000044
dtype: float64
probability of word low: 0    0.000044
dtype: float64
probability of word sister: 0    0.000533
dtype: float64
probability of word encouraged: 0    0.000044
dtype: float64
probability of word griefs;: 0    0.000044
dtype: float64
probability of word loyal: 0    0.000089
dtype: float64
probability of word always--: 0    0.000089
dtype: float64
probability of word there--: 0    0.000089
dtype: float64
probability of word cousin: 0    0.000044
dtype: float64
probability of word brother-germain: 0    0.000089
dtype: float64
probability of word alliance: 0    0.000089
dtype: float64
probability of word [10]: 0    0.000089
dtype: float64
probability of word [11]: 0    0.000089
dtype: float64
probability of word friendship--: 0    0.000089
dtype: float64
probability of wo

dtype: float64
probability of word 1e1: 0    0.000133
dtype: float64
probability of word 1e9: 0    0.000178
dtype: float64
probability of word posted: 0    0.000044
dtype: float64
probability of word distribution: 0    0.000044
dtype: float64
probability of word additional: 0    0.000044
dtype: float64
probability of word linked: 0    0.000044
dtype: float64
probability of word binary: 0    0.000044
dtype: float64
probability of word version: 0    0.000044
dtype: float64
probability of word (wwwgutenbergorg): 0    0.000044
dtype: float64
probability of word displaying: 0    0.000044
dtype: float64
probability of word providing: 0    0.000044
dtype: float64
probability of word provided: 0    0.000044
dtype: float64
probability of word payments: 0    0.000044
dtype: float64
probability of word tax: 0    0.000044
dtype: float64
probability of word notifies: 0    0.000044
dtype: float64
probability of word s/he: 0    0.000044
dtype: float64
probability of word medium: 0    0.000044
dtype: 

In [8]:
my_new_dict1 = pd.DataFrame.from_dict(new_dict1, orient='index')
words_probabilities1={}
for word in new_dict1:            
    probab = new_dict1[word]/ my_new_dict1.sum()
    words_probabilities1[word]=probab
    print('probability of word '+ word +': ' + str(probab))

probability of word mask: 0    0.001338
dtype: float64
probability of word dumas: 0    0.001338
dtype: float64
probability of word introduction: 0    0.001338
dtype: float64
probability of word the: 0    0.038796
dtype: float64
probability of word playwright: 0    0.000669
dtype: float64
probability of word had: 0    0.002676
dtype: float64
probability of word a: 0    0.01204
dtype: float64
probability of word adventures: 0    0.000669
dtype: float64
probability of word almost: 0    0.001338
dtype: float64
probability of word and: 0    0.019398
dtype: float64
probability of word readers: 0    0.000669
dtype: float64
probability of word friends: 0    0.00602
dtype: float64
probability of word english: 0    0.001338
dtype: float64
probability of word history: 0    0.002676
dtype: float64
probability of word form: 0    0.000669
dtype: float64
probability of word brief: 0    0.000669
dtype: float64
probability of word novels: 0    0.001338
dtype: float64
probability of word 1625: 0    0.00

dtype: float64
probability of word regulations: 0    0.000669
dtype: float64
probability of word final: 0    0.000669
dtype: float64
probability of word historical: 0    0.000669
dtype: float64
probability of word compare: 0    0.000669
dtype: float64
probability of word enjoy: 0    0.001338
dtype: float64
probability of word prisoner: 0    0.008696
dtype: float64
probability of word order: 0    0.000669
dtype: float64
probability of word place: 0    0.000669
dtype: float64
probability of word was: 0    0.010033
dtype: float64
probability of word said: 0    0.004013
dtype: float64
probability of word merely: 0    0.000669
dtype: float64
probability of word resounded: 0    0.000669
dtype: float64
probability of word towers: 0    0.000669
dtype: float64
probability of word luxury: 0    0.000669
dtype: float64
probability of word effected: 0    0.000669
dtype: float64
probability of word curious: 0    0.000669
dtype: float64
probability of word were: 0    0.001338
dtype: float64
probabili

dtype: float64
probability of word spent--: 0    0.001338
dtype: float64
probability of word aspect: 0    0.001338
dtype: float64
probability of word still: 0    0.001338
dtype: float64
probability of word candor;: 0    0.000669
dtype: float64
probability of word intrenched: 0    0.000669
dtype: float64
probability of word think;: 0    0.000669
dtype: float64
probability of word nothing--nothing: 0    0.000669
dtype: float64
probability of word so: 0    0.005351
dtype: float64
probability of word impatience: 0    0.000669
dtype: float64
probability of word you’: 0    0.001338
dtype: float64
probability of word noisy-le-sec: 0    0.000669
dtype: float64
probability of word flame-colored: 0    0.000669
dtype: float64
probability of word hair: 0    0.001338
dtype: float64
probability of word astonished: 0    0.000669
dtype: float64
probability of word xiii’s: 0    0.000669
dtype: float64
probability of word musketeers: 0    0.001338
dtype: float64
probability of word which: 0    0.004

dtype: float64
probability of word freedom: 0    0.001338
dtype: float64
probability of word thought: 0    0.001338
dtype: float64
probability of word weary: 0    0.001338
dtype: float64
probability of word itself: 0    0.000669
dtype: float64
probability of word speak: 0    0.001338
dtype: float64
probability of word nor: 0    0.001338
dtype: float64
probability of word mirrors: 0    0.001338
dtype: float64
probability of word mine: 0    0.000669
dtype: float64
probability of word here: 0    0.002007
dtype: float64
probability of word precaution: 0    0.001338
dtype: float64
probability of word end: 0    0.001338
dtype: float64
probability of word instructed: 0    0.000669
dtype: float64
probability of word st: 0    0.000669
dtype: float64
probability of word iv: 0    0.002676
dtype: float64
probability of word nearly: 0    0.001338
dtype: float64
probability of word facts: 0    0.000669
dtype: float64
probability of word mansion: 0    0.000669
dtype: float64
probability of word hopes

dtype: float64


In [9]:
my_new_dict2 = pd.DataFrame.from_dict(new_dict2, orient='index')
words_probabilities2={}
for word in new_dict2:            
    probab = new_dict2[word]/ my_new_dict2.sum()
    words_probabilities2[word]=probab
    print('probability of word '+ word +': ' + str(probab))

probability of word notice: 0    0.002597
dtype: float64
probability of word worthy: 0    0.002597
dtype: float64
probability of word gentleman: 0    0.005195
dtype: float64
probability of word were: 0    0.002597
dtype: float64
probability of word king: 0    0.002597
dtype: float64
probability of word intended: 0    0.002597
dtype: float64
probability of word in: 0    0.005195
dtype: float64
probability of word witnessed: 0    0.002597
dtype: float64
probability of word d’artagnan: 0    0.018182
dtype: float64
probability of word about: 0    0.002597
dtype: float64
probability of word a: 0    0.005195
dtype: float64
probability of word him: 0    0.005195
dtype: float64
probability of word pensive--nay: 0    0.002597
dtype: float64
probability of word only: 0    0.005195
dtype: float64
probability of word host: 0    0.002597
dtype: float64
probability of word slashes: 0    0.002597
dtype: float64
probability of word and: 0    0.01039
dtype: float64
probability of word entrance: 0    

dtype: float64
probability of word mine: 0    0.002597
dtype: float64
probability of word pistoles: 0    0.005195
dtype: float64
probability of word myself: 0    0.005195
dtype: float64
probability of word skirt: 0    0.002597
dtype: float64
probability of word knee: 0    0.005195
dtype: float64
probability of word half: 0    0.007792
dtype: float64
probability of word always: 0    0.002597
dtype: float64
probability of word made: 0    0.002597
dtype: float64
probability of word month: 0    0.005195
dtype: float64
probability of word contrary: 0    0.005195
dtype: float64
probability of word last: 0    0.002597
dtype: float64
probability of word size: 0    0.005195
dtype: float64
probability of word on: 0    0.002597
dtype: float64
probability of word been: 0    0.002597
dtype: float64
probability of word court: 0    0.005195
dtype: float64
probability of word now: 0    0.002597
dtype: float64
probability of word tailor: 0    0.012987
dtype: float64
probability of word suit: 0    0.005

Thansforming our disctionaries into DataFrames to facilitate our job later on.

In [10]:
words_probabilities = pd.DataFrame.from_dict(words_probabilities, orient='index')
words_probabilities1 = pd.DataFrame.from_dict(words_probabilities1, orient='index')
words_probabilities2 = pd.DataFrame.from_dict(words_probabilities2, orient='index')

# Hidden Markov Model

Hidden Markov Model (HMM) is a statistical Markov model in which the system being modeled is assumed to be a Markov process with unobserved (i.e. hidden) states.

## Mathematical description

A basic hidden Markov model can be described as follows:

$T$ = length of the observation sequence

$N$ = number of states in the model

$M$ = number of observations symbols

$Q$ = ${q_0, q_1, \ldots, q_{N-1} }$ = distinct states of the Markov process

$V$ = ${0, 1, \ldots, M-1 }$ = set of possible observations

$A$ = state transition probabilities

$B$ = observation probability matrix

$\pi$ = initial state distribution

$O$ = $(o_0, O_1, \ldots, o_{T-1})$ = observation sequence

Now let's define our hidden states. We also define our starting probabilities, i.e., the initial state disreibution $$ \pi $$

In [11]:
np.random.seed(20190326)

hidden_states = ['chap1', 'chap2']
pi = [0.5, 0.5]
state_space = pd.Series(pi, index=hidden_states, name='states')
print(state_space)
print('\n', state_space.sum())

chap1    0.5
chap2    0.5
Name: states, dtype: float64

 1.0


Now we need to define the state transition probabilities, i.e., the matrix **A** above.

In [12]:
a_df = pd.DataFrame(columns=hidden_states, index=hidden_states)
a_df.loc[hidden_states[0]] = [0.7, 0.3]
a_df.loc[hidden_states[1]] = [0.4, 0.6]

Let's check our state transition probabilities matrix **A** :

In [13]:
print(a_df)

a = a_df.values
print('\n', a, a.shape, '\n')
print(a_df.sum(axis=1))

      chap1 chap2
chap1   0.7   0.3
chap2   0.4   0.6

 [[0.7 0.3]
 [0.4 0.6]] (2, 2) 

chap1    1.0
chap2    1.0
dtype: float64


As expected the probabilities for each hidden state sum to 1.

Let's update the new_dict1 with the data from new_dict2. This way we will have in one single new_dict1 a list of the words that appears in both dictionaries (our hidden states).

In [14]:
new_dict1.update(new_dict2)
observable_states = new_dict1.keys()

Checking:

In [15]:
new_dict1

{'1625': 1,
 '_ennui_': 1,
 '_fetes_': 1,
 '_me_': 1,
 '_too_': 1,
 'a': 2,
 'about': 1,
 'absence': 1,
 'accident': 1,
 'act': 2,
 'adamant': 2,
 'adventures': 1,
 'affairs': 1,
 'affection': 1,
 'affrighted': 1,
 'after': 2,
 'again': 2,
 'ago': 2,
 'ah': 1,
 'alas': 2,
 'all': 2,
 'almost': 2,
 'along': 2,
 'already': 1,
 'also': 1,
 'alteration': 2,
 'always': 1,
 'am': 2,
 'ambitious': 1,
 'an': 1,
 'and': 4,
 'angles': 1,
 'announcing': 1,
 'answer': 4,
 'antiquity': 2,
 'any': 4,
 'aramis': 2,
 'architects': 1,
 'are': 1,
 'are--': 2,
 'armchair': 3,
 'arms': 2,
 'arrest': 1,
 'arriving': 1,
 'ask': 1,
 'asked': 3,
 'asleep': 1,
 'aspect': 2,
 'assassinated': 2,
 'assuredly': 1,
 'astonished': 1,
 'at': 7,
 'athos': 1,
 'attached': 1,
 'attachment': 1,
 'attention': 3,
 'attitude': 2,
 'austria': 2,
 'austria--': 2,
 'aware': 1,
 'back': 1,
 'baisemeaux': 2,
 'barrier': 1,
 'bars': 2,
 'base;': 1,
 'bastile': 9,
 'battle': 1,
 'be': 2,
 'become': 1,
 'been': 1,
 'before': 3,
 'b

Now create another 2 disctionaries populating with the probabilities of word occurence in each hidden state.

In [16]:
words_probabilities_1 ={}
for word in new_dict:
    if not word in words_probabilities1.index.values:
        words_probabilities_1[word] = 0
    else:
        words_probabilities_1[word] = words_probabilities1.loc[word, 0]

words_probabilities_1 = pd.DataFrame.from_dict(words_probabilities_1, orient='index')

words_probabilities_2 ={}
for word in new_dict:
    if not word in words_probabilities2.index.values:
        words_probabilities_2[word] = 0
    else:
        words_probabilities_2[word] = words_probabilities2.loc[word, 0]

words_probabilities_2 = pd.DataFrame.from_dict(words_probabilities_2, orient='index')

Checking both dictionaries:

* words_probabilities_1
* words_probabilities_2

In [17]:
print("Shape: %s" % str(words_probabilities_1.shape))
print("------------------------------------------------------------------")
print("Column names: %s" % str(words_probabilities_1.columns))
print("------------------------------------------------------------------")
str(words_probabilities_1.info())
print("------------------------------------------------------------------")
print(words_probabilities_1.describe())
print("------------------------------------------------------------------")
words_probabilities_1.head()

Shape: (3978, 1)
------------------------------------------------------------------
Column names: RangeIndex(start=0, stop=1, step=1)
------------------------------------------------------------------
<class 'pandas.core.frame.DataFrame'>
Index: 3978 entries, mask to ebooks
Data columns (total 1 columns):
0    3978 non-null float64
dtypes: float64(1)
memory usage: 46.6+ KB
------------------------------------------------------------------
                 0
count  3978.000000
mean      0.000251
std       0.001216
min       0.000000
25%       0.000000
50%       0.000000
75%       0.000000
max       0.038796
------------------------------------------------------------------


0
mask          0.001338
dumas         0.001338
introduction  0.001338
the           0.038796
playwright    0.000669

In [18]:
print("Shape: %s" % str(words_probabilities_2.shape))
print("------------------------------------------------------------------")
print("Column names: %s" % str(words_probabilities_2.columns))
print("------------------------------------------------------------------")
str(words_probabilities_2.info())
print("------------------------------------------------------------------")
print(words_probabilities_2.describe())
print("------------------------------------------------------------------")
words_probabilities_2.head()

Shape: (3978, 1)
------------------------------------------------------------------
Column names: RangeIndex(start=0, stop=1, step=1)
------------------------------------------------------------------
<class 'pandas.core.frame.DataFrame'>
Index: 3978 entries, mask to ebooks
Data columns (total 1 columns):
0    3978 non-null float64
dtypes: float64(1)
memory usage: 46.6+ KB
------------------------------------------------------------------
                 0
count  3978.000000
mean      0.000251
std       0.001508
min       0.000000
25%       0.000000
50%       0.000000
75%       0.000000
max       0.033766
------------------------------------------------------------------


0
mask          0.000000
dumas         0.000000
introduction  0.000000
the           0.020779
playwright    0.000000

Having both dictionaries (latter transformed to DataFrames) allow us to define the matrix **B**, i.e., the observation probability matrix.

In [19]:
b_df = pd.DataFrame(columns=observable_states, index=hidden_states)
b_df.loc[hidden_states[0]] = words_probabilities_1.loc[:,0]
b_df.loc[hidden_states[1]] = words_probabilities_2.loc[:,0]

Let's check our state transition probabilities matrix **B** :

In [20]:
print(b_df)
b = b_df.values
print('\n', b, b.shape, '\n')
print(b_df.sum(axis=1))


             mask       dumas introduction        the   playwright  \
chap1  0.00133779  0.00133779   0.00133779   0.038796  0.000668896   
chap2           0           0            0  0.0207792            0   

              had           a   adventures      almost        and    ...      \
chap1  0.00267559   0.0120401  0.000668896  0.00133779   0.019398    ...       
chap2           0  0.00519481            0           0  0.0103896    ...       

       invitation  following   list--’   something   percerin          m  \
chap1           0          0           0           0          0          0   
chap2  0.00519481  0.0025974  0.00519481  0.00519481  0.0103896  0.0025974   

             busy      said;      worse       _me_  
chap1           0          0          0          0  
chap2  0.00519481  0.0025974  0.0025974  0.0025974  

[2 rows x 699 columns]

 [[0.0013377926421404682 0.0013377926421404682 0.0013377926421404682 ...,
  0.0 0.0 0.0]
 [0.0 0.0 0.0 ..., 0.0025974025974025974

We see that we have 699 observations for both hidden states.
As expected, each row, the sum of all 699 observatios for each hidden state adds to 1.

## Multinomial HMM 

Since we have all the information needed, this allows us to  our Hidden Markov Model. We opted to use a Multinomial HMM.

In [21]:
model = hmm.MultinomialHMM(n_components=2)
model.startprob_ = np.array(pi)
model.transmat_ = a_df.astype(float)
model.emissionprob_ = b_df.astype(float) 

Let's check what are probabilities for the first 4 observations in our matrix **B**

Recall that they are:

In [22]:
b_df.columns[0:4]

Index(['mask', 'dumas', 'introduction', 'the'], dtype='object')

For 'mask'

In [23]:
math.exp(model.score(np.array([[0]])))


0.0006688963210702338

For 'dumas'

In [24]:
math.exp(model.score(np.array([[1]])))

0.0006688963210702338

For 'introduction'

In [25]:
math.exp(model.score(np.array([[2]])))

0.0006688963210702338

For 'the'

In [26]:
math.exp(model.score(np.array([[3]])))

0.02978760370064718

Now let's sample a chain of 5, or 10, or 25 observations from our Multinomial HMM

In [27]:
X5, Y5 = model.sample(5)

for x in X5:
    print(b_df.columns[x])

Index(['despair'], dtype='object')
Index(['water'], dtype='object')
Index(['during'], dtype='object')
Index(['do'], dtype='object')
Index(['vallon--'], dtype='object')


The chain of hidden states from the sample above are:

In [28]:
print(Y5)

[1 1 0 1 1]


In [29]:
X10, Y10 = model.sample(10)

for x in X10:
    print(b_df.columns[x])

Index(['himself'], dtype='object')
Index(['continued'], dtype='object')
Index(['period'], dtype='object')
Index(['me'], dtype='object')
Index(['numerous'], dtype='object')
Index(['thin'], dtype='object')
Index(['the'], dtype='object')
Index(['again'], dtype='object')
Index(['occasion'], dtype='object')
Index(['humbly'], dtype='object')


The chain of hidden states from the sample above are:

In [30]:
print(Y10)

[0 0 1 1 0 1 0 1 1 1]


In [31]:
X25, Y25 = model.sample(25)

for x in X25:
    print(b_df.columns[x])

Index(['garden;'], dtype='object')
Index(['on'], dtype='object')
Index(['the'], dtype='object')
Index(['monseigneur'], dtype='object')
Index(['enemy'], dtype='object')
Index(['for'], dtype='object')
Index(['tremble'], dtype='object')
Index(['heaven'], dtype='object')
Index(['it'], dtype='object')
Index(['the'], dtype='object')
Index(['again'], dtype='object')
Index(['scarcely'], dtype='object')
Index(['know'], dtype='object')
Index(['have'], dtype='object')
Index(['you'], dtype='object')
Index(['and'], dtype='object')
Index(['baisemeaux'], dtype='object')
Index(['paris'], dtype='object')
Index(['year--'], dtype='object')
Index(['from'], dtype='object')
Index(['louis'], dtype='object')
Index(['aramis'], dtype='object')
Index(['freedom'], dtype='object')
Index(['gentleman'], dtype='object')
Index(['eyes'], dtype='object')


The chain of hidden states from the sample above are:

In [32]:
print(Y25)

[0 1 1 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 1 1]


Based on the last three sample chain, with 5, 10 and 25 observations respectively, we can predict the hidden states:

In [33]:
model.predict(X5)

array([1, 1, 0, 1, 1])

In [34]:
model.predict(X10)

array([0, 0, 1, 1, 0, 1, 1, 1, 1, 1])

In [35]:
model.predict(X25)

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       1, 1])